cli/flox-rust-sdk/src/providers/flox_cpp_utils.rs

```diff
@@ -0,0 +1,281 @@
+
+use enum_dispatch::enum_dispatch;
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+use tracing::debug;
+
+use crate::models::pkgdb::{
+    call_pkgdb,
+    error_codes,
+    CallPkgDbError,
+    ContextMsgError,
+    PkgDbError,
+    PKGDB_BIN,
+};
+
+#[derive(Debug, Error)]
+pub enum FlakeInstallableError {
+    #[error(transparent)]
+    Pkgdb(#[from] CallPkgDbError),
+    // todo: do we need to break this into more specific errors?
+    #[error("Failed to lock flake installable: {0}")]
+    LockInstallable(String),
+    #[error("Failed to deserialize locked installable")]
+    DeserializeLockedInstallable(#[from] serde_json::Error),
+}
+
+/// Rust representation of the output of `pkgdb lock-flake-installable`
+/// This is a direct translation of the definition in
+/// `<flox>/pkgdb/include/flox/lock-flake-installable.hh`
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+#[serde(rename_all = "kebab-case")]
+#[cfg_attr(test, derive(proptest_derive::Arbitrary))]
+pub struct LockedInstallable {
+    /// locked url of the flakeref component of the installable
+    pub(crate) locked_url: String,
+    pub(crate) flake_description: Option<String>,
+    pub(crate) locked_flake_attr_path: String,
+    pub(crate) derivation: String,
+    /// Map of output names to their paths
+    /// The values are expected to be nix store paths
+    pub(crate) outputs: BTreeMap<String, String>,
+    /// List of output names in the original order
+    pub(crate) output_names: Vec<String>,
+    /// List of output names to install as defined by the package
+    pub(crate) outputs_to_install: Option<Vec<String>>,
+    /// List of output names to install as requested by the user
+    pub(crate) requested_outputs_to_install: Option<Vec<String>>,
+    /// System as defined by the package
+    pub(crate) package_system: String,
+    /// System as defined by the specified target system
+    pub(crate) locked_system: String,
+    pub(crate) name: String,
+    pub(crate) pname: Option<String>,
+    pub(crate) version: Option<String>,
+    pub(crate) description: Option<String>,
+    pub(crate) licenses: Option<Vec<String>>,
+    pub(crate) broken: Option<bool>,
+    pub(crate) unfree: Option<bool>,
+}
+
+/// Required functionality to lock a flake installable
+///
+/// Implemented as a trait to allow mocking as evaluation is
+/// time-consuming unless cached.
+///
+/// The trait is implemented by the [`Pkgdb`] struct which is the canonical implementation
+/// using the `pkgdb lock-flake-installable` command.
+///
+/// The trait is also implemented by the [`LockFlakeInstallableMock`] struct which is used for testing.
+#[enum_dispatch]
+pub trait LockFlakeInstallableTrait {
+    fn lock_flake_installable(
+        &self,
+        system: impl AsRef<str>,
+        installable: impl AsRef<str>,
+    ) -> Result<LockedInstallable, FlakeInstallableError>;
+}
+
+#[derive(Debug)]
+#[enum_dispatch(LockFlakeInstallableTrait)]
+pub enum LockFlakeInstallable {
```

Comment:
Suggestion: again the name is a little confusing, this enum sounds like it's describing the thing to be locked rather than the object doing the locking.